### ISSUE TITLE
Stack overflow for primary boot thread on ARM64

### ISSUE THREAD

**Issue by b49020** [OP-TEE/optee_os#3064]
With latest upstream `optee_os`, stack canary check for primary boot thread fails on Developerbox as follows:

```
I/TC: Initialized
D/TC:00 0 init_primary_helper:1104 Primary CPU switching to normal world boot
E/TC:00   Dead canary at start of 'stack_tmp[0]'
E/TC:00   Panic at core/arch/arm/kernel/thread.c:194 <thread_check_canaries>
E/TC:00   Call stack:
E/TC:00    0x00000000fc00cb84
```

Bisection leads to following commit (Thanks to @FulcronZ for the report and bisection).

```
    commit ad565116a0d7664e57df14a82d87f7dc54b1fb28
    Author: Jerome Forissier <jerome.forissier@linaro.org>
    Date:   Mon Feb 25 17:01:46 2019 +0100

        core: crypto: libtomcrypt: enable LTC_CLEAN_STACK

        Enables LTC_CLEAN_STACK so that LibTomCrypt will wipe key material and
        other sensitive data once no longer used.
```

It sounds like a generic issue on ARM64 where allocated stack isn't sufficient for primary boot thread. But it has only surfaced after `libtomcrypt` explicitly clears its stack usage (leading to stack canary corruption) after above patch. So we might need to consider increasing stack allocation for primary boot thread as following patch resolves the issue:

```
diff --git a/core/arch/arm/kernel/thread.c b/core/arch/arm/kernel/thread.c
index 2b14dcf..ba32de4 100644
--- a/core/arch/arm/kernel/thread.c
+++ b/core/arch/arm/kernel/thread.c
@@ -61,7 +61,7 @@
 #endif /*ARM32*/
 
 #ifdef ARM64
-#define STACK_TMP_SIZE         (2048 + STACK_TMP_OFFS)
+#define STACK_TMP_SIZE         (3072 + STACK_TMP_OFFS)
 #define STACK_THREAD_SIZE      8192
 
 #if TRACE_LEVEL > 0
```

**Comment by b49020** [comment]
> Increasing the TMP stack is quite expensive, that's why I'd like to double check that it's the only sensible way forward.

Agree.

It seems that simple SHA512 computation on Developerbox causes this stack overflow issue. Following is analysis that I have done:

Patch applied to dump stack addresses:

```
static int  sha512_compress(hash_state * md, unsigned char *buf)
 {
     ulong64 S[8], W[80], t0, t1;
     int i;
+    DMSG("i addr: %p, S: %p, W: %p, t0: %p, t1: %p", &i, S, W, &t0, &t1);
<snip>
 int sha512_done(hash_state * md, unsigned char *out)
 {
     int i;
+    DMSG("i addr: %p", &i);
<snip>
 void burn_stack(unsigned long len)
 {
    unsigned char buf[32];
+   DMSG("buf addr: %p", buf);
```

Execution dump:

```
D/TC:00 0 init_canaries:173 #Stack canaries for stack_tmp[0] with top at 0xfc0b6638
D/TC:00 0 init_canaries:173 watch *0xfc0b663c
<snip>
D/TC:00   sha512_done:239 i addr: 0xfc0b63f4
D/TC:00   _sha512_compress:130 i addr: 0xfc0b60cc, S: 0xfc0b60e0, W: 0xfc0b6120, t0: 0xfc0b60d0, t1: 0xfc0b60d8
D/TC:00   burn_stack:55 buf addr: 0xfc0b6380
D/TC:00   burn_stack:55 buf addr: 0xfc0b6340
D/TC:00   burn_stack:55 buf addr: 0xfc0b6300
D/TC:00   burn_stack:55 buf addr: 0xfc0b62c0
D/TC:00   burn_stack:55 buf addr: 0xfc0b6280
D/TC:00   burn_stack:55 buf addr: 0xfc0b6240
D/TC:00   burn_stack:55 buf addr: 0xfc0b6200
D/TC:00   burn_stack:55 buf addr: 0xfc0b61c0
D/TC:00   burn_stack:55 buf addr: 0xfc0b6180
D/TC:00   burn_stack:55 buf addr: 0xfc0b6140
D/TC:00   burn_stack:55 buf addr: 0xfc0b6100
D/TC:00   burn_stack:55 buf addr: 0xfc0b60c0
D/TC:00   burn_stack:55 buf addr: 0xfc0b6080
D/TC:00   burn_stack:55 buf addr: 0xfc0b6040
D/TC:00   burn_stack:55 buf addr: 0xfc0b6000
D/TC:00   burn_stack:55 buf addr: 0xfc0b5fc0
D/TC:00   burn_stack:55 buf addr: 0xfc0b5f80
D/TC:00   burn_stack:55 buf addr: 0xfc0b5f40
D/TC:00   burn_stack:55 buf addr: 0xfc0b5f00
D/TC:00   burn_stack:55 buf addr: 0xfc0b5ec0
D/TC:00   burn_stack:55 buf addr: 0xfc0b5e80
D/TC:00   burn_stack:55 buf addr: 0xfc0b5e40
D/TC:00   burn_stack:55 buf addr: 0xfc0b5e00
```

As observed from above dump, `burn_stack()` cleans stack upto `0xfc0b5e00` which is `2104` bytes less than stack top: `0xfc0b6638`. So this leads to stack canary corruption.

But looking carefully at `_sha512_compress()` api dump, it only uses stack upto `0xfc0b60cc` i.e. `1388` bytes less than stack top: `0xfc0b6638`. This shows that `burn_stack()` is somewhat broken as it cleans that stack space too which isn't used (approx. `726` bytes in this case).

After carefully observation of `burn_stack()` api, it seems that it doesn't take into account the meta-data pushed on stack on every function call (like lr, fp etc) but assumes only `buf` is placed onto stack in each recursive call. So actual recursive call takes `64` bytes of stack (as visible from dump) but below code assumes it as `32` bytes only.

```
/**
   Burn some stack memory
   @param len amount of stack to burn in bytes
*/
void burn_stack(unsigned long len)
{
   unsigned char buf[32];
   DMSG("buf addr: %p", buf);
   zeromem(buf, sizeof(buf));
   if (len > (unsigned long)sizeof(buf))
      burn_stack(len - sizeof(buf));
}
```

So to summarize above findings, it seems like usage of stack space is still under `2048` bytes but logic of clearing stack in libtomcrypt needs to be fixed.

Any suggestions/comments are welcome.

**Comment by jforissier** [comment]
IMHO, this `burn_stack()` method is FUBAR. How can one assume anything about what stack space has been consumed by the compiler for a given function call? Think about padding, inlining, and probably other things...

**Comment by b49020** [comment]
Yeah, proper way would be to clear stack in corresponding function only. But that would require changes in imported `libtomcrypt` code. Should we report this issue upstream and get it fixed there first?

**Comment by b49020** [comment]
Have a look at description here [1] when `burn_stack()` was included in `libtomcrypt`.

[1] https://github.com/libtom/libtomcrypt/blob/3fb0eea01b181a2616407a36eadb67af3b964ffb/changes#L1283

**Comment by jenswi-linaro** [comment]
We could report it straight away and also just fix the code. Cleaning should be done with `memzero_explicit()` in each function as you described it @b49020. When we have a proper solution we could try to upstream that (possibly using `zeromem()` instead of `memzero_explicit()`).

**Comment by b49020** [comment]
Issue reported here [1]. Will send a patch to use `memzero_explicit()` instead of `burn_stack()` api usage.

[1] https://github.com/libtom/libtomcrypt/issues/486

**Comment by sjaeckel** [comment]
> Will send a patch to use `memzero_explicit()` instead of `burn_stack()` api usage.

I'm looking forward to your solution of this problem

**Comment by b49020** [comment]
@sjaeckel Thanks for your acknowledgement of the problem. I have tried to address the problem here [1]. I would be happy to get any feedback from you.

[1] libtom/libtomcrypt#486

**Comment by b49020** [comment]
@jenswi-linaro How about if we just make `zeromem()` a wrapper to call `memzero_explicit()`? It will make least changes to code imported from `LibTomCrypt`.

**Comment by jenswi-linaro** [comment]
@b49020, sounds like a good idea.

**Comment by b49020** [comment]
@jenswi-linaro Can we accept that variable length array fix (proposed here [1]) for `burn_stack` api for 3.6.0 release until we reach on an agreement in upstream? It seems to involve a minimal change.

[1] libtom/libtomcrypt#486

**Comment by jenswi-linaro** [comment]
Fine with me, it fixes a real error.
